### PR TITLE
Support adding attachment on bug update

### PR DESCRIPTION
**Pull request by cproensa**
Add attachments section in bug update page and change status page.

Fixes: #21819, #21820

### CODE REVIEW

File: bug_change_status_page.php [hunk 1]
```diff
@@ -386,6 +387,33 @@
 <?php
 	}
 
+	if( $t_allow_file_upload ) {
```
Comment (atrol) [inline bug_change_status_page.php]: After merging as is, we would have nearly the same code block four times (bug_change_status_page.php, bug_report_page.php, bug_update_page.php, bugnote_add_inc.php)
Maybe you can introduce some kind of `function print_bugnote_attach_zone( ... )`

**Comment by cproensa** [comment]
> If I copy an image, I can't paste it in the new added places to add attachments, but can paste into the existing ones.

Fixed

> If I just add an attachment from existing places, I can hover over their left column later and can make them private, delete, etc. For the ones added via your code, I can only delete them. This is likely being caused by such attachments not having an associated empty note.

Fixed
However, note that currently, attachments added at report time also don't have a related note and can't be made private, only delete.

**Comment by dregad** [comment]
@vboctor @atrol can you please confirm you're OK with the latest fixes following your respetive reviews ?

**Comment by atrol** [comment]
I don't see any change related to my remark concerning code duplication.
Didn't test, but I assume there will also be two more places where we encounter https://www.mantisbt.org/bugs/view.php?id=26627

**Comment by cproensa** [comment]
> I don't see any change related to my remark concerning code duplication.

It's in my to-do list

**Comment by vboctor** [comment]
> @cproensa However, note that currently, attachments added at report time also don't have a related note and can't be made private, only delete.

This is by design since the attachments issued at report time is associated with the issue itself and it derives its private/public from the issue and in the future, it may be displayed within the issue itself.  If the user wants to report a public issue with private attachments, then it is a two step process (report an issue, then attach files to it privately).

File: bug_update.php [hunk 2]
```diff
@@ -412,11 +412,25 @@
...
+	'time_tracking' => array(
+		'duration' => $t_bug_note->time_tracking
+	),
+	'files' => helper_array_transpose( $f_files )
+);
+$t_note_data = array(
+	'query' => $t_note_query,
+	'payload' => $t_note_payload,
+);
+$t_note_command = new IssueNoteAddCommand( $t_note_data );
```
Comment (vboctor) [inline bug_update.php]: The commands should reflect the intention of the user and not be used as re-usable code.  For example, the user intent it to change status or update issue in this case.  Using command here may trigger emails, events, and other semantics that is not expected.  The #1400 PR was attempting at having an update command.

Anyway, whether we have an update command or not, the re-usable code between add note and update issue should be in APIs that are called by both commands.

Comment (cproensa) [reply bug_update.php]: If not using the command, we are at the previous situation: file api don't provide a way to add an attachment that satisfies the expected internal representation.

**Comment by cproensa** [comment]
> I don't see any change related to my remark concerning code duplication.

I have rewritten the form contents, so now it's a separated div:
- visually groups the note specific fields, so they are not confused with the fields of the main action form.
- reusable
- the note section can be expanded/collapsed for a leaner presentation

> Didn't test, but I assume there will also be two more places where we encounter https://www.mantisbt.org/bugs/view.php?id=26627

that's fixed now.

**Comment by cproensa** [comment]
> The commands should reflect the intention of the user and not be used as re-usable code. For example, the user intent it to change status or update issue in this case. Using command here may trigger emails, events, and other semantics that is not expected. The #1400 PR was attempting at having an update command.

I'm not sure that applies for the "add note" command. Adding a note is a differentiated enough action that can be done along other action (eg, update of the issue).
Do you consider that adding a note is part of "updating the isse"? Then, why have a separate note command, when it should be all managed in a "proper" issue update command?
Regarding notifications, in the current implementation for notifications, i'd rather have a notification for issue updated, and another one for the note added. Because, they are formatted differently, and if i expect notes to be notified in the short format, i may overlook a note that was added along an update since there is not specific information that the note was added in that kind of notification.
Additionally, currently its not working very good, eg: https://mantisbt.org/bugs/view.php?id=26701

**Comment by dregad** [comment]
Came back to this following @atrol's note in #1631. 

Not sure what is the status on this, @vboctor @cproensa ? Can it be merged or are you still discussing ?

**Comment by cproensa** [comment]
>  Can it be merged or are you still discussing ?
No merge yet.
My questions are still open, regarding the command, but not blocking though.
There is an UI change that needs some confirmation.
If the UI normalization is valid, it can probably be applied to some other note forms (eg, group update, reminder...)

**Comment by vboctor** [comment]
@cproensa sorry for the delay.  Had another look.

-  I'm OK we the UI update to have notes collapsed by default, but let's see how users find it.
- Can you also update the Edit Note form to support attaching files?
- For the use of `IssueNoteAddCommand` command, we can use it but I wouldn't trigger a different email for the note.  This can be done by adding an "options" block similar to "query" and "payload" with a flag to disable email notification `email_notification`.  The default would be ON, but you can override.  The options block is designed to pass such internal flags that we don't want to be usable via REST API.  You can checkout `IssueAddCommand` which uses options.

**Comment by vboctor** [comment]
@cproensa any chance of finalizing this PR to merge it?

**Comment by cproensa** [comment]
> @cproensa any chance of finalizing this PR to merge it?

Would like to, but at the moment i have little time, and it's complicated with the current situation.I guess in some weeks i can get back to it.
I hope you all are doing well, take care.

**Comment by dregad** [comment]
@cproensa weeks have turned into months, any chance to see this finalized ?

**Comment by aaribaud** [comment]
Hello, I am interested in this PR. @cproensa, do you plan to update it?

**Comment by vboctor** [comment]
Closing stale PR.

**Pull request closed without merging**